### PR TITLE
Upgrade interview agent

### PR DESCRIPTION
**Pull request by simonryrie**
- Upgrades to use Assistants API and new chat GPT model
- Enables retrieval functionality with knowledge of attached pdf

### CODE REVIEW

**Comment by eliazzo** [comment]
In utils/parsePdfToText.ts you have switched to ```require``` vs the ```import``` statements used in app/api/generate/route.ts

Is it a good idea to stick to one?

**Comment by simonryrie** [comment]
> In utils/parsePdfToText.ts you have switched to `require` vs the `import` statements used in app/api/generate/route.ts
> 
> Is it a good idea to stick to one?

Good spot, had loads of trouble with this! It's because when using fs it has to be server side using Node so could only get it to work using require syntax, however this meant I couldn't use it dynamically as I hoped (right now it can only be used to send the parsed text to a text file rather than pass the value into the route.ts automatically).

We actually shouldn't need this function any more though because now the API is upgraded to assistant it can retrieve from pdfs directly. I kept it in because the retrieval is slow in its current form - I thought it might be faster for it to read text so we might consider parsing manually first, but right now it's not being used.

**Comment by eliazzo** [comment]
I see! I guess that's why you set the type:module in the package.json? This meant that I was unable to run your branch on my local because the next config file is not defined in ES module scope. I can get past this by changing the next config and postcss config files to .cjs but I think if we're not going to use the parsePdfToText we might want to revert the package.json (remove the type:module) instead of renaming files. Did you have any issues with the config files after changing the package.json?

**Comment by eliazzo** [comment]
> > In utils/parsePdfToText.ts you have switched to `require` vs the `import` statements used in app/api/generate/route.ts
> > Is it a good idea to stick to one?
> 
> Good spot, had loads of trouble with this! It's because when using fs it has to be server side using Node so could only get it to work using require syntax, however this meant I couldn't use it dynamically as I hoped (right now it can only be used to send the parsed text to a text file rather than pass the value into the route.ts automatically).
> 
> We actually shouldn't need this function any more though because now the API is upgraded to assistant it can retrieve from pdfs directly. I kept it in because the retrieval is slow in its current form - I thought it might be faster for it to read text so we might consider parsing manually first, but right now it's not being used.

I see! I guess that's why you set the type:module in the package.json? This meant that I was unable to run your branch on my local because the next config file is not defined in ES module scope. I can get past this by changing the next config and postcss config files to .cjs but I think if we're not going to use the parsePdfToText we might want to revert the package.json (remove the type:module) instead of renaming files. Did you have any issues with the config files after changing the package.json? @simonryrie

**Comment by simonryrie** [comment]
> > > In utils/parsePdfToText.ts you have switched to `require` vs the `import` statements used in app/api/generate/route.ts
> > > Is it a good idea to stick to one?
> > 
> > 
> > Good spot, had loads of trouble with this! It's because when using fs it has to be server side using Node so could only get it to work using require syntax, however this meant I couldn't use it dynamically as I hoped (right now it can only be used to send the parsed text to a text file rather than pass the value into the route.ts automatically).
> > We actually shouldn't need this function any more though because now the API is upgraded to assistant it can retrieve from pdfs directly. I kept it in because the retrieval is slow in its current form - I thought it might be faster for it to read text so we might consider parsing manually first, but right now it's not being used.
> 
> I see! I guess that's why you set the type:module in the package.json? This meant that I was unable to run your branch on my local because the next config file is not defined in ES module scope. I can get past this by changing the next config and postcss config files to .cjs but I think if we're not going to use the parsePdfToText we might want to revert the package.json (remove the type:module) instead of renaming files. Did you have any issues with the config files after changing the package.json? @simonryrie

I actually can't remember why type: modules was there but it was stopping that function working when I tried to run it again and wasn't needed for the main functionality so removed it now!

I also tested using the parsed text file for retrieval and it wasn't any faster so have deleted parsePdfToText anyway.

I think it might slow because it's running the whole assistant creation / file upload / retrieval from scratch on every post request - I believe we can do this once only but need to find a way to persist the data across multiple exchanges.

If you try now hopefully the main chatbot functionality should work and have knowledge of the assessment criteria, just very slowly @eliazzo

**Review by eliazzo — COMMENTED** [review]
Before merging shall we
- Uninstall supabase and fs
- Delete generatePrompt.ts as it is empty
- Delete the pdf file

**Comment by simonryrie** [comment]
> Before merging shall we
> 
> * Uninstall supabase and fs
> * Delete generatePrompt.ts as it is empty
> * Delete the pdf file

Had one last clear up, should be ready now! 
- Deleted everything supabase-related
- generatePrompt.ts was already deleted for me, hopefully is for you as well now?
- We still need fs and the pdf file to upload the file for now until we can share the same Open AI key / assistant - even then probably good to keep them because we will need them whenever we create a new assistant.

File: app/api/create-assistant/route.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    // File and assistant setup only needs to run once, then comment out and get assistant id from Open AI account
+    // We can share an assistant once we are on the same Open AI API key
+
+    const file = await openai.files.create({
+      file: fs.createReadStream("data/SDL4.pdf"),
+      purpose: "assistants",
+    });
+
+    const assistant = await openai.beta.assistants.create({
+      instructions: `You are helping a software developer apprentice write a portfolio for their training programme, ask them questions about their job and projects they've worked on. Don't go into too much detail, you only need a basic overview of these. When you have a basic overview of their job and their work projects, say goodbye and end the conversation. Do not mention the file uploaded unless asked, or read it unless you need to - it provides a detailed description of the assessment criteria for the apprenticeship.`,
```
Comment (cameochoquer) [inline app/api/create-assistant/route.ts:31]: should this be 'end point assessment' or training programme? I don't think it matters at the moment but for future refinement might be good to narrow it does or clarify it?

File: app/api/submit-message/route.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      runStatus = await openai.beta.threads.runs.retrieve(threadId, run.id);
+      // Sleep for a short time before checking again
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+    } while (runStatus.status !== "completed");
+
+    const messages = await openai.beta.threads.messages.list(threadId);
+
+    const formattedMessages = messages.data.map((msg) => ({
+      role: msg.role,
+      content: msg.content[0].text.value,
```
Comment (cameochoquer) [inline app/api/submit-message/route.ts:43]: we made need to address this as there is a false positive typescript error
see from issue notes below: 

_To be direct, this isn't an incomplete type – this is TypeScript being pedantic about the fact that an element of the content array could either be {type: 'text', text: string} or {type: 'image_url', image_url: string} and if it's the latter, the text property won't exist.
You need to first check whether type === 'text' before accessing .text.
In other words, while this can be toilsome, it is working as intended._

`  const messages = data.map((message) =>
    message?.content.map((content) => (content.type === 'text' ? content.text : '')).join(''),
  );`
